src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/security/authentication/MontaguAuthenticator.kt

```diff
@@ -0,0 +1,62 @@
+package org.vaccineimpact.orderlyweb.security.authentication
+
+import org.pac4j.core.context.WebContext
+import org.pac4j.core.credentials.TokenCredentials
+import org.pac4j.core.credentials.authenticator.Authenticator
+import org.pac4j.core.exception.CredentialsException
+import org.pac4j.core.profile.CommonProfile
+import org.pac4j.core.util.CommonHelper
+import org.slf4j.LoggerFactory
+import org.vaccineimpact.orderlyweb.db.UserRepository
+import org.vaccineimpact.orderlyweb.models.UserSource
+
+class MontaguAuthenticator(private val userRepository: UserRepository,
+                           private val montaguClient: MontaguAPIClient
+) : Authenticator<TokenCredentials>
+{
+    private val logger = LoggerFactory.getLogger(MontaguAuthenticator::class.java)
```

Comment:
This logging code looks to be identical between the two authenticators, could go in a base class (which would need to find its subclass at runtime to pass to the factory)? But see my other comment - it feels like errors ought to get logged automatically. And there will be other types of objects which will want to log errors and warnings with minimal code, so I wonder if each of them looking after their own logger object is going to feel a bit noisy eventually. Could just have a global logging service which any object can call in one line e.g. Logging. log(this, WARN, "message");

Reply:
Yes good points.

In fact I realised that we do already log all errors in the errorhandler (see e.g. https://vimc.myjetbrains.com/youtrack/agiles/103-2/104-86?issue=VIMC-2802)  (only drawback is they're reported as being thrown by ` org.vaccineimpact.orderlyweb.app_start.ErrorHandler` rather than the class where they originated.)